streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -595,6 +595,72 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
       ZStream.fromQueue(queue).unTake
     }
 
+  /**
+   * Creates a stream that passes its elements through a queue that drops elements
+   * (dropping or sliding). It implements a promise that signals abrupt completion
+   * of the stream due to an error or the end of the stream. After the promise has
+   * been completed, it continues to offer the current (at the time) contents of the
+   * queue, until it has been exhausted, at which point it finishes with the signal
+   * from the promise.
+   */
+  private final def bufferSignal[E1 >: E, A1 >: A](queue: Queue[Take[E1, A1]]): ZStream[R, E1, A1] =
+    ZStream.managed {
+      for {
+        signal <- Promise.make[Cause[E], Unit].toManaged_
+        _ <- self
+              .foreachManaged(a => queue.offer(Take.Value(a)).unit)
+              .foldCauseM(
+                c => signal.fail(c).unit.toManaged_,
+                _ => signal.succeed(()).unit.toManaged_
+              )
+              .fork
+      } yield (queue, signal)
+    }.flatMap {
+      case (queue, signal) =>
+        def pull: UIO[Take[E1, A1]] =
```

Comment:
Very clever! My only comment here is that this is a busy-wait when the queue is empty and the stream has not completed yet.

Have you considered implementing this using two queues? One for the signal and one for the data? Then you can race the takes.

Reply:
I'm aware yes, that's why I wanted your input on this. I'll rework it. Thanks.

Reply:
Ok 👍 If you do end up using this approach, use `raceWith` and keep the losing fiber in a ref, don't cancel it. I think there are some edge conditions where if you cancel a `queue.take`, you might lose the element.

Reply:
What do I do with the ref in the end?

Reply:
You re-race the fiber's `join` against `data.take` until the fiber succeeds, then you wait for the queue to empty as you've done here

Reply:
Done.